src/wallet/rpcwallet.cpp

```diff
@@ -2556,18 +2555,9 @@ static UniValue loadwallet(const JSONRPCRequest& request)
         }
     }
 
-    std::string warning;
-    if (!CWallet::Verify(*g_rpc_interfaces->chain, location, false, error, warning)) {
-        throw JSONRPCError(RPC_WALLET_ERROR, "Wallet file verification failed: " + error);
-    }
-
-    std::shared_ptr<CWallet> const wallet = CWallet::CreateWalletFromFile(*g_rpc_interfaces->chain, location);
-    if (!wallet) {
-        throw JSONRPCError(RPC_WALLET_ERROR, "Wallet loading failed.");
-    }
-    AddWallet(wallet);
-
-    wallet->postInitProcess();
+    std::string error, warning;
+    std::shared_ptr<CWallet> const wallet = LoadWallet(*g_rpc_interfaces->chain, location, error, warning);
```

Comment:
Why isn't the file existence checking above also moved to `LoadWallet()`? If I call `LoadWallet()` with a filename that doesn't exist or doesn't contain a wallet.dat file, `LoadWallet()` will create a new wallet file. I don't think that's desired behaviour.

I can trigger this by hovering over the 'Open Wallet' menu in qt, then deleting the wallet file in a different terminal, then trying to load the wallet.

Reply:
Correct me if I'm wrong but technically a race would still exist even if the check is moved, it would be hard to trigger it.

Anyway, in order to move the check to `LoadWallet` I have to somehow map the "not found error" to `RPC_WALLET_NOT_FOUND` - at the moment if `LoadWallet` fails the raised error is `RPC_WALLET_ERROR` - so I would have to change the interface.

Doing what you suggest has the advantage of reducing references to `wallet.dat`:
https://github.com/bitcoin/bitcoin/blob/1bc149d05b09d716723d2f091250fab38fd70fc2/src/wallet/rpcwallet.cpp#L2564
And leave it just to src/wallet/db.cpp.

Considering I'd like to refactor `interfaces::Node::loadWallet` to be asynchronous - to know load progress, to cancel/abort, to know errors/warns - I'll have to change these calls too.

Considering the above ACK's I suggest to follow up your and @Sjors improvements.

WDYT?

Reply:
> technically a race would still exist even if the check is moved

Yes, you're right. The race is still there, but it's very hard to trigger if the checks are moved into `LoadWallet()`.

Maybe this is too hacky, but perhaps the error code raised by the RPC could be set based on the error string returned by `LoadWallet()`? Or we could just change the return code in this case to a generic `RPC_WALLET_ERROR`.

Whatever way we go, I think it's fine to defer changing this to a future PR.

Reply:
That's why I think we should just try to open instead of checking before opening - those checks could be removed.

Reply:
> we should just try to open instead of checking before opening

Sure, as long as 'try to open' doesn't create a new wallet if one isn't there (which I believe is what happens currently).